File: packages/hushspec/tests/receipt.test.ts
import { describe, expect, it } from 'vitest';
import type { HushSpec } from '../src/schema.js';
import type { EvaluationAction } from '../src/evaluate.js';
import { evaluate } from '../src/evaluate.js';
import { HUSHSPEC_VERSION } from '../src/version.js';
import {
  evaluateAudited,
  computePolicyHash,
  DEFAULT_AUDIT_CONFIG,
} from '../src/receipt.js';
import type { AuditConfig, DecisionReceipt } from '../src/receipt.js';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;
const ISO_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/;
const SHA256_RE = /^[0-9a-f]{64}$/;

function minimalSpec(): HushSpec {
  return {
    hushspec: '0.1.0',
    name: 'test-policy',
  };
}

function specWithToolAccess(): HushSpec {
  return {
    hushspec: '0.1.0',
    name: 'tool-policy',
    rules: {
      tool_access: {
        allow: ['read_file', 'write_file'],
        block: ['dangerous_tool'],
        default: 'block',
      },
    },
  };
}

function enabledConfig(): AuditConfig {
  return { ...DEFAULT_AUDIT_CONFIG };
}

function disabledConfig(): AuditConfig {
  return { enabled: false, include_rule_trace: false, redact_content: true };
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe('evaluateAudited', () => {
  it('returns correct decision matching evaluate()', () => {
    const spec = specWithToolAccess();
    const action: EvaluationAction = { type: 'tool_call', target: 'read_file' };
    const receipt = evaluateAudited(spec, action, enabledConfig());
    const result = evaluate(spec, action);

    expect(receipt.decision).toBe(result.decision);
    expect(receipt.decision).toBe('allow');
  });

  it('returns deny for blocked tool matching evaluate()', () => {
    const spec = specWithToolAccess();
    const action: EvaluationAction = { type: 'tool_call', target: 'dangerous_tool' };
    const receipt = evaluateAudited(spec, action, enabledConfig());
    const result = evaluate(spec, action);

    expect(receipt.decision).toBe(result.decision);
    expect(receipt.decision).toBe('deny');
    expect(receipt.matched_rule).toBe(result.matched_rule);
  });

  it('has valid UUID receipt_id', () => {
    const spec = minimalSpec();
    const action: EvaluationAction = { type: 'tool_call', target: 'test' };
    const receipt = evaluateAudited(spec, action, enabledConfig());

    expect(receipt.receipt_id).toMatch(UUID_RE);
  });

  it('has valid ISO timestamp', () => {
    const spec = minimalSpec();
    const action: EvaluationAction = { type: 'tool_call', target: 'test' };
    const receipt = evaluateAudited(spec, action, enabledConfig());

    expect(receipt.timestamp).toMatch(ISO_RE);
  });

  it('sets hushspec_version to current version', () => {
    const spec = minimalSpec();
    const action: EvaluationAction = { type: 'tool_call', target: 'test' };
    const receipt = evaluateAudited(spec, action, enabledConfig());

    expect(receipt.hushspec_version).toBe(HUSHSPEC_VERSION);
  });

  it('populates rule trace when enabled', () => {
    const spec = specWithToolAccess();
    const action: EvaluationAction = { type: 'tool_call', target: 'read_file' };
    const receipt = evaluateAudited(spec, action, enabledConfig());

    expect(receipt.rule_trace.length).toBeGreaterThan(0);
    expect(receipt.rule_trace[0].rule_block).toBe('tool_access');
    expect(receipt.rule_trace[0].evaluated).toBe(true);
  });

  it('returns empty trace when config disabled', () => {
    const spec = specWithToolAccess();
    const action: EvaluationAction = { type: 'tool_call', target: 'read_file' };
    const receipt = evaluateAudited(spec, action, disabledConfig());

    expect(receipt.rule_trace).toEqual([]);
    expect(receipt.evaluation_duration_us).toBe(0);
  });

  it('returns empty policy hash when config disabled', () => {
    const spec = specWithToolAccess();
    const action: EvaluationAction = { type: 'tool_call', target: 'read_file' };
    const receipt = evaluateAudited(spec, action, disabledConfig());

    expect(receipt.policy.content_hash).toBe('');
  });

  it('sets content_redacted when content present and redact enabled', () => {
    const spec: HushSpec = {
      hushspec: '0.1.0',
      rules: {
        shell_commands: {
          enabled: true,
          forbidden_patterns: [],
        },
      },
    };
    const action: EvaluationAction = {
      type: 'shell_command',
      target: 'echo hello',
      content: 'some content here',
    };
    const receipt = evaluateAudited(spec, action, enabledConfig());

    expect(receipt.action.content_redacted).toBe(true);
  });

  it('does not set content_redacted when no content', () => {
    const spec = minimalSpec();
    const action: EvaluationAction = { type: 'tool_call', target: 'test' };
    const receipt = evaluateAudited(spec, action, enabledConfig());

    expect(receipt.action.content_redacted).toBe(false);
  });

  it('has non-negative evaluation_duration_us when enabled', () => {
    const spec = specWithToolAccess();
    const action: EvaluationAction = { type: 'tool_call', target: 'read_file' };
    const receipt = evaluateAudited(spec, action, enabledConfig());

    expect(receipt.evaluation_duration_us).toBeGreaterThanOrEqual(0);
  });

  it('generates unique receipt IDs', () => {
    const spec = minimalSpec();
    const action: EvaluationAction = { type: 'tool_call', target: 'test' };

    const receipt1 = evaluateAudited(spec, action, enabledConfig());
    const receipt2 = evaluateAudited(spec, action, enabledConfig());

    expect(receipt1.receipt_id).not.toBe(receipt2.receipt_id);
  });

  it('includes action type and target in summary', () => {
    const spec = minimalSpec();
    const action: EvaluationAction = { type: 'egress', target: 'api.example.com' };
    const receipt = evaluateAudited(spec, action, enabledConfig());

    expect(receipt.action.type).toBe('egress');
    expect(receipt.action.target).toBe('api.example.com');
  });

  it('populates policy name from spec', () => {
    const spec = minimalSpec();
    const action: EvaluationAction = { type: 'tool_call', target: 'test' };
    const receipt = evaluateAudited(spec, action, enabledConfig());

    expect(receipt.policy.name).toBe('test-policy');
    expect(receipt.policy.version).toBe('0.1.0');
  });
});

describe('computePolicyHash', () => {
  it('produces a valid SHA-256 hex string', () => {
    const spec = minimalSpec();
    const hash = computePolicyHash(spec);

    expect(hash).toMatch(SHA256_RE);
  });

  it('is deterministic', () => {
    const spec = minimalSpec();
    const hash1 = computePolicyHash(spec);
    const hash2 = computePolicyHash(spec);

    expect(hash1).toBe(hash2);
  });

  it('differs for different specs', () => {
    const spec1 = minimalSpec();
    const spec2: HushSpec = {
      hushspec: '0.1.0',
      name: 'different-policy',
    };

    expect(computePolicyHash(spec1)).not.toBe(computePolicyHash(spec2));
  });
});

describe('rule trace for different action types', () => {
  it('traces egress rule', () => {
    const spec: HushSpec = {
      hushspec: '0.1.0',
      rules: {
        egress: {
          allow: ['api.example.com'],
          default: 'block',
        },
      },
    };
    const action: EvaluationAction = { type: 'egress', target: 'api.example.com' };
    const receipt = evaluateAudited(spec, action, enabledConfig());

    expect(receipt.decision).toBe('allow');
    const egressTrace = receipt.rule_trace.find(t => t.rule_block === 'egress');
    expect(egressTrace).toBeDefined();
    expect(egressTrace!.evaluated).toBe(true);
    expect(egressTrace!.outcome).toBe('allow');
  });

  it('traces shell_commands rule', () => {
    const spec: HushSpec = {
      hushspec: '0.1.0',
      rules: {
        shell_commands: {
          enabled: true,
          forbidden_patterns: ['rm\\s+-rf'],
        },
      },
    };
    const action: EvaluationAction = { type: 'shell_command', target: 'ls -la' };
    const receipt = evaluateAudited(spec, action, enabledConfig());

    expect(receipt.decision).toBe('allow');
    const shellTrace = receipt.rule_trace.find(t => t.rule_block === 'shell_commands');
    expect(shellTrace).toBeDefined();
    expect(shellTrace!.evaluated).toBe(true);
    expect(shellTrace!.outcome).toBe('allow');
  });

  it('traces skip for unconfigured tool_access', () => {
    const spec: HushSpec = { hushspec: '0.1.0' };
    const action: EvaluationAction = { type: 'tool_call', target: 'test' };
    const receipt = evaluateAudited(spec, action, enabledConfig());

    const toolTrace = receipt.rule_trace.find(t => t.rule_block === 'tool_access');
    expect(toolTrace).toBeDefined();
    expect(toolTrace!.evaluated).toBe(false);
    expect(toolTrace!.outcome).toBe('skip');
  });

  it('handles unknown action type with default trace', () => {
    const spec: HushSpec = { hushspec: '0.1.0' };
    const action: EvaluationAction = { type: 'unknown_action', target: 'test' };
    const receipt = evaluateAudited(spec, action, enabledConfig());

    expect(receipt.decision).toBe('allow');
    const defaultTrace = receipt.rule_trace.find(t => t.rule_block === 'default');
    expect(defaultTrace).toBeDefined();
    expect(defaultTrace!.evaluated).toBe(true);
  });
});

File: packages/go/hushspec/receipt_test.go
package hushspec

import (
	"regexp"
	"testing"
)

var (
	uuidRE   = regexp.MustCompile(`^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`)
	isoRE    = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}`)
	sha256RE = regexp.MustCompile(`^[0-9a-f]{64}$`)
)

func minimalSpec() *HushSpec {
	return &HushSpec{
		HushSpecVersion: "0.1.0",
		Name:            "test-policy",
	}
}

func specWithToolAccess() *HushSpec {
	return &HushSpec{
		HushSpecVersion: "0.1.0",
		Name:            "tool-policy",
		Rules: &Rules{
			ToolAccess: &ToolAccessRule{
				Enabled: true,
				Allow:   []string{"read_file", "write_file"},
				Block:   []string{"dangerous_tool"},
				Default: DefaultActionBlock,
			},
		},
	}
}

func enabledConfig() *AuditConfig {
	c := DefaultAuditConfig()
	return &c
}

func disabledConfig() *AuditConfig {
	return &AuditConfig{
		Enabled:          false,
		IncludeRuleTrace: false,
		RedactContent:    true,
	}
}

func TestEvaluateAuditedDecisionParity(t *testing.T) {
	spec := specWithToolAccess()
	action := &EvaluationAction{Type: "tool_call", Target: "read_file"}
	receipt := EvaluateAudited(spec, action, enabledConfig())
	result := Evaluate(spec, action)

	if receipt.Decision != result.Decision {
		t.Errorf("decision mismatch: receipt=%q, evaluate=%q", receipt.Decision, result.Decision)
	}
	if receipt.Decision != DecisionAllow {
		t.Errorf("expected allow, got %q", receipt.Decision)
	}
}

func TestEvaluateAuditedDenyParity(t *testing.T) {
	spec := specWithToolAccess()
	action := &EvaluationAction{Type: "tool_call", Target: "dangerous_tool"}
	receipt := EvaluateAudited(spec, action, enabledConfig())
	result := Evaluate(spec, action)

	if receipt.Decision != result.Decision {
		t.Errorf("decision mismatch: receipt=%q, evaluate=%q", receipt.Decision, result.Decision)
	}
	if receipt.Decision != DecisionDeny {
		t.Errorf("expected deny, got %q", receipt.Decision)
	}
	if receipt.MatchedRule != result.MatchedRule {
		t.Errorf("matched_rule mismatch: receipt=%q, evaluate=%q", receipt.MatchedRule, result.MatchedRule)
	}
}

func TestReceiptHasValidUUID(t *testing.T) {
	spec := minimalSpec()
	action := &EvaluationAction{Type: "tool_call", Target: "test"}
	receipt := EvaluateAudited(spec, action, enabledConfig())

	if !uuidRE.MatchString(receipt.ReceiptID) {
		t.Errorf("receipt_id %q does not match UUID pattern", receipt.ReceiptID)
	}
}

func TestReceiptHasValidTimestamp(t *testing.T) {
	spec := minimalSpec()
	action := &EvaluationAction{Type: "tool_call", Target: "test"}
	receipt := EvaluateAudited(spec, action, enabledConfig())

	if !isoRE.MatchString(receipt.Timestamp) {
		t.Errorf("timestamp %q does not match ISO 8601 pattern", receipt.Timestamp)
	}
}

func TestReceiptSetsHushSpecVersion(t *testing.T) {
	spec := minimalSpec()
	action := &EvaluationAction{Type: "tool_call", Target: "test"}
	receipt := EvaluateAudited(spec, action, enabledConfig())

	if receipt.HushSpecVersion != Version {
		t.Errorf("expected hushspec_version %q, got %q", Version, receipt.HushSpecVersion)
	}
}

func TestRuleTracePopulatedWhenEnabled(t *testing.T) {
	spec := specWithToolAccess()
	action := &EvaluationAction{Type: "tool_call", Target: "read_file"}
	receipt := EvaluateAudited(spec, action, enabledConfig())

	if len(receipt.RuleTrace) == 0 {
		t.Fatal("expected non-empty rule trace")
	}
	if receipt.RuleTrace[0].RuleBlock != "tool_access" {
		t.Errorf("expected first trace block to be tool_access, got %q", receipt.RuleTrace[0].RuleBlock)
	}
	if !receipt.RuleTrace[0].Evaluated {
		t.Error("expected first trace to be evaluated")
	}
}

func TestEmptyTraceWhenDisabled(t *testing.T) {
	spec := specWithToolAccess()
	action := &EvaluationAction{Type: "tool_call", Target: "read_file"}
	receipt := EvaluateAudited(spec, action, disabledConfig())

	if len(receipt.RuleTrace) != 0 {
		t.Errorf("expected empty trace, got %d entries", len(receipt.RuleTrace))
	}
	if receipt.EvaluationDurationUs != 0 {
		t.Errorf("expected zero duration, got %d", receipt.EvaluationDurationUs)
	}
}

func TestEmptyPolicyHashWhenDisabled(t *testing.T) {
	spec := specWithToolAccess()
	action := &EvaluationAction{Type: "tool_call", Target: "read_file"}
	receipt := EvaluateAudited(spec, action, disabledConfig())

	if receipt.Policy.ContentHash != "" {
		t.Errorf("expected empty content_hash, got %q", receipt.Policy.ContentHash)
	}
}

func TestContentRedactedWhenContentPresent(t *testing.T) {
	spec := &HushSpec{
		HushSpecVersion: "0.1.0",
		Rules: &Rules{
			ShellCommands: &ShellCommandsRule{
				Enabled:           true,
				ForbiddenPatterns: []string{},
			},
		},
	}
	action := &EvaluationAction{
		Type:    "shell_command",
		Target:  "echo hello",
		Content: "some content here",
	}
	receipt := EvaluateAudited(spec, action, enabledConfig())

	if !receipt.Action.ContentRedacted {
		t.Error("expected content_redacted to be true")
	}
}

func TestContentNotRedactedWhenNoContent(t *testing.T) {
	spec := minimalSpec()
	action := &EvaluationAction{Type: "tool_call", Target: "test"}
	receipt := EvaluateAudited(spec, action, enabledConfig())

	if receipt.Action.ContentRedacted {
		t.Error("expected content_redacted to be false")
	}
}

func TestNonNegativeDurationWhenEnabled(t *testing.T) {
	spec := specWithToolAccess()
	action := &EvaluationAction{Type: "tool_call", Target: "read_file"}
	receipt := EvaluateAudited(spec, action, enabledConfig())

	if receipt.EvaluationDurationUs < 0 {
		t.Errorf("expected non-negative duration, got %d", receipt.EvaluationDurationUs)
	}
}

func TestUniqueReceiptIDs(t *testing.T) {
	spec := minimalSpec()
	action := &EvaluationAction{Type: "tool_call", Target: "test"}
	r1 := EvaluateAudited(spec, action, enabledConfig())
	r2 := EvaluateAudited(spec, action, enabledConfig())

	if r1.ReceiptID == r2.ReceiptID {
		t.Error("expected unique receipt IDs")
	}
}

func TestActionSummaryFields(t *testing.T) {
	spec := minimalSpec()
	action := &EvaluationAction{Type: "egress", Target: "api.example.com"}
	receipt := EvaluateAudited(spec, action, enabledConfig())

	if receipt.Action.Type != "egress" {
		t.Errorf("expected action type egress, got %q", receipt.Action.Type)
	}
	if receipt.Action.Target != "api.example.com" {
		t.Errorf("expected target api.example.com, got %q", receipt.Action.Target)
	}
}

func TestPolicyNameFromSpec(t *testing.T) {
	spec := minimalSpec()
	action := &EvaluationAction{Type: "tool_call", Target: "test"}
	receipt := EvaluateAudited(spec, action, enabledConfig())

	if receipt.Policy.Name != "test-policy" {
		t.Errorf("expected policy name test-policy, got %q", receipt.Policy.Name)
	}
	if receipt.Policy.Version != "0.1.0" {
		t.Errorf("expected policy version 0.1.0, got %q", receipt.Policy.Version)
	}
}

func TestComputePolicyHashValid(t *testing.T) {
	spec := minimalSpec()
	hash := ComputePolicyHash(spec)

	if !sha256RE.MatchString(hash) {
		t.Errorf("hash %q does not match SHA-256 pattern", hash)
	}
}

func TestComputePolicyHashDeterministic(t *testing.T) {
	spec := minimalSpec()
	h1 := ComputePolicyHash(spec)
	h2 := ComputePolicyHash(spec)

	if h1 != h2 {
		t.Errorf("hash not deterministic: %q vs %q", h1, h2)
	}
}

func TestComputePolicyHashDiffersForDifferentSpecs(t *testing.T) {
	spec1 := minimalSpec()
	spec2 := &HushSpec{
		HushSpecVersion: "0.1.0",
		Name:            "different-policy",
	}

	if ComputePolicyHash(spec1) == ComputePolicyHash(spec2) {
		t.Error("expected different hashes for different specs")
	}
}

func TestTraceEgressRule(t *testing.T) {
	spec := &HushSpec{
		HushSpecVersion: "0.1.0",
		Rules: &Rules{
			Egress: &EgressRule{
				Enabled: true,
				Allow:   []string{"api.example.com"},
				Default: DefaultActionBlock,
			},
		},
	}
	action := &EvaluationAction{Type: "egress", Target: "api.example.com"}
	receipt := EvaluateAudited(spec, action, enabledConfig())

	if receipt.Decision != DecisionAllow {
		t.Errorf("expected allow, got %q", receipt.Decision)
	}

	found := false
	for _, tr := range receipt.RuleTrace {
		if tr.RuleBlock == "egress" {
			found = true
			if !tr.Evaluated {
				t.Error("expected egress trace to be evaluated")
			}
			if tr.Outcome != RuleOutcomeAllow {
				t.Errorf("expected allow outcome, got %q", tr.Outcome)
			}
		}
	}
	if !found {
		t.Error("no egress trace entry found")
	}
}

func TestTraceShellCommands(t *testing.T) {
	spec := &HushSpec{
		HushSpecVersion: "0.1.0",
		Rules: &Rules{
			ShellCommands: &ShellCommandsRule{
				Enabled:           true,
				ForbiddenPatterns: []string{`rm\s+-rf`},
			},
		},
	}
	action := &EvaluationAction{Type: "shell_command", Target: "ls -la"}
	receipt := EvaluateAudited(spec, action, enabledConfig())

	if receipt.Decision != DecisionAllow {
		t.Errorf("expected allow, got %q", receipt.Decision)
	}

	found := false
	for _, tr := range receipt.RuleTrace {
		if tr.RuleBlock == "shell_commands" {
			found = true
			if !tr.Evaluated {
				t.Error("expected shell_commands trace to be evaluated")
			}
			if tr.Outcome != RuleOutcomeAllow {
				t.Errorf("expected allow outcome, got %q", tr.Outcome)
			}
		}
	}
	if !found {
		t.Error("no shell_commands trace entry found")
	}
}

func TestTraceSkipUnconfiguredToolAccess(t *testing.T) {
	spec := &HushSpec{HushSpecVersion: "0.1.0"}
	action := &EvaluationAction{Type: "tool_call", Target: "test"}
	receipt := EvaluateAudited(spec, action, enabledConfig())

	found := false
	for _, tr := range receipt.RuleTrace {
		if tr.RuleBlock == "tool_access" {
			found = true
			if tr.Evaluated {
				t.Error("expected tool_access trace to not be evaluated")
			}
			if tr.Outcome != RuleOutcomeSkip {
				t.Errorf("expected skip outcome, got %q", tr.Outcome)
			}
		}
	}
	if !found {
		t.Error("no tool_access trace entry found")
	}
}

func TestTraceUnknownActionType(t *testing.T) {
	spec := &HushSpec{HushSpecVersion: "0.1.0"}
	action := &EvaluationAction{Type: "unknown_action", Target: "test"}
	receipt := EvaluateAudited(spec, action, enabledConfig())

	if receipt.Decision != DecisionAllow {
		t.Errorf("expected allow, got %q", receipt.Decision)
	}

	found := false
	for _, tr := range receipt.RuleTrace {
		if tr.RuleBlock == "default" {
			found = true
			if !tr.Evaluated {
				t.Error("expected default trace to be evaluated")
			}
		}
	}
	if !found {
		t.Error("no default trace entry found")
	}
}
